### PR TITLE
feat(tty): add cbreak option to enable passthrough of signals

### PR DESCRIPTION
**Pull request by marcushultman**
Closes https://github.com/denoland/deno/issues/8343.

### CODE REVIEW

**Comment by ry** [comment]
Hi @marcushultman - thanks for the patch - and sorry for the slow review.

I certainly see the need for exposing this API, but I'm a bit hesitant to invent new names like `generateSignal`. I wonder if you could do some research into how this API looks in other platforms (Node, Rust, Go, etc). Maybe we can decide on a name that others are already using.

Also we need a test somehow... I think it's fine to leave this unimplemented in Windows - but it should probably throw some NotImplemented exception if people try to use it there.

**Comment by marcushultman** [comment]
Thanks @ry! Glad to hear that there may be a way to get this available.

As far as research goes,  [Node](https://nodejs.org/api/tty.html#tty_readstream_setrawmode_mode) states that "all special processing of characters by the terminal is disabled". Golang also has a [raw mode](https://godoc.org/golang.org/x/crypto/ssh/terminal#MakeRaw), but that unconditionally captures ctrl+c just like Node (this is not documented).

Deno just wraps [termios](https://man7.org/linux/man-pages/man3/termios.3.html), so it would be reasonable to align with naming from there. In termios, they call it "local modes" when the `c_lflag` bitfield is set. The `ISIG` flag doesn't really state a specific "mode":
> When any of the characters INTR, QUIT, SUSP, or DSUSP are re‐
              ceived, generate the corresponding signal.  
This is where I chose the `generateSignal` name from.

Perhaps `ISIG`, `isigMode`, `interruptSignal` or similar could be used?

As for testing, I could not easily think of a way to raise ^C in the current test suite(s). Not sure what can be done instead. Throwing NotImplemented on windows however should be doable!

**Comment by ry** [comment]
Regarding the naming, I want to summon some experts for advice. @piscisaureus, @bnoordhuis What do you think about this `generatesSignal` name? Do you have a better suggestion?

> As for testing, I could not easily think of a way to raise ^C in the current test suite(s). Not sure what can be done instead. Throwing NotImplemented on windows however should be doable!

I think you'll want to create a custom test in cli/tests/integration_tests.rs and create a pty. You can probably spawn a Deno subprocess (using std::process::Command) and send the SIGINT signal using kill(2) I think...

Have a look at https://github.com/denoland/deno/blob/91f293442d5bfb28212b5b13b536ceae06427c99/cli/tests/integration_tests.rs#L174-L222

File: cli/dts/lib.deno.unstable.d.ts [hunk 1]
```diff
@@ -815,19 +815,28 @@ declare namespace Deno {
...
    * Set TTY to be under raw mode or not. In raw mode, characters are read and
    * returned as is, without being processed. All special processing of
    * characters by the terminal is disabled, including echoing input characters.
+   * Characters that correspond to a signal can optionally be enabled.
    * Reading from a TTY device in raw mode is faster than reading from a TTY
    * device in canonical mode.
    *
    * ```ts
-   * Deno.setRaw(myTTY.rid, true);
+   * Deno.setRaw(myTTY.rid, true, { generateSignal: true });
```
Comment (bnoordhuis) [inline cli/dts/lib.deno.unstable.d.ts:832]: `generateSignal` only does something when `mode == true`, right? That should probably be made clearer.

**Review by bnoordhuis — COMMENTED** [review]
Apropos the name, I'd not overthink it and just call it `cbreak`.

People might reasonably expect `generateSignals` to cover `SIGTTOU` but `ISIG` doesn't, that's the `TOSTOP` flag.

**Comment by marcushultman** [comment]
> Apropos the name, I'd not overthink it and just call it `cbreak`.

That would work.

Indeed, this flag is only relevant when when going into raw mode (or cbreak mode actually). What is your preferred way to model this? Documentation only/throw exception on invalid input or make a new API endpoint all together? Maybe that question is for @ry

**Comment by marcushultman** [comment]
@ry The existing test for `setRaw` (`test_raw_tty`) Is deactivated and broken in more ways than the linked issue suggest. I ran in to the same behavior when trying to write a test for this change. My strategy was to `await Deno.signals.interrupt()` in the typescript test, and `master.write_all(&[3])` from the test case side - this prints "^C" to the TTY, which in some way is expected, but considering that we wouldn't be able to have the test running because of https://github.com/denoland/deno/issues/4262, I didn't see a reason to include that test.

**Comment by ry** [comment]
@marcushultman Maybe we can just skip the test on mac? Or are there deeper issues? Let me try to start another PR which re-enables that test except on mac.

**Comment by ry** [comment]
@marcushultman I've landed a fix to test_raw_tty. I think it should be working on mac now too.

**Comment by marcushultman** [comment]
@ry that's great! I'll try again to get a test passing for this

**Comment by marcushultman** [comment]
@ry The test uses `pty::fork` where the child process spawns deno and shares stdio/stdout. When writing SIGINT to the master process, this is not propagated to the deno process (but it can be captured in the child process). I don't know how to properly trigger `Deno.signals.interrupt()` for a deno process launched with `Command::new`. Ideas are welcome!

**Review by ry — APPROVED** [review]
Thanks @marcushultman - I made a small change to your test but otherwise LGTM

**Pull request merged**